Allow coin-selection to carry an extra "reserve input"

KtorZ opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->

#1821 

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- 859f53bdeed9fc641f83a98201e184254eb14591
  :round_pushpin: **extend 'CoinSelection' type to carry an extra reserve amount that can be used as input**
  For now, the reserve is always 'Nothing', which should not alter any of the existing behavior. Then, it can be
set to a specific value (e.g. the reward balance) to be used on the left side of the balance (i.e. the input side).

- eebc981a6e80d9403f0df5c9dc5a9917cf5289e0
  :round_pushpin: **rename 'rebalanceChangeOutput' in 'rebalanceSelection' and allow picking from the reserve**
  The approach is still very 'simple', before trying to deplete any change output, we first try to remove fees
from the reserve. Once the reserve is empty, we'll start depleting change outputs.

- 08a6cbe852793398b642d70faeae834d64620406
  :round_pushpin: **move prop_rebalanceSelection from byron package to core package**
  It was originally in `-byron` because it was using the fee policy and
fee estimation from Byron, so, while moving it, I also changed the way
the transaction size is estimated to make it mimics the way it's done on
shelley and byron, with rather realistic fee values.

- 338dde73e990924e3004d4e3a11825deb0382a2e
  :round_pushpin: **generate arbitrary selection with reserve, add more classification and assertions to properties**
  
- 20c161bd8cb528d1c6a16cfada568dcebc5dd137
  :round_pushpin: **do not mutate the reserve as part of the fee balancing process**
  Everything in the past coin selection (without reserve) was easy to reason about. One could always look at the sum
of inputs and compare it with the sum of outputs and changes, and there were some nice properties about this. Having
mutable reserve makes it somewhat very confusing because, there's no trace of where the money comes from (turning
reserve into changes shouldn't make it disappear).

- 22e17144e678bd0292acc3891c186af2fa0d4cc4
  :round_pushpin: **expand 'reserve' to also cover deposit and reclaim**
  So that types are clear from the coin selection and a bit less confusing.

- 43cace27cff105b6d3a8428bbec2998ada0398a0
  :round_pushpin: **extend transaction layer with an 'initDelegationSelection'**
    This in order to defer to the relevant backend the necessary bits of logic regarding withdrawal, deposits and reclaims.
  I've also removed the 'WalletDelegation' from 'mkDelegationJoinTx' as this can now be inferred from the presence of a deposit in the coin selection.

  We have the following situation:

  object     | cardano-node | jormungandr
   ----      | ----         | ----
  inputs     | explicit     | explicit
  reclaim    | implicit     | N/A
  withdrawal | explicit     | N/A
  ---        | ---          | ---
  outputs    | explicit     | explicit
  deposit    | implicit     | N/A

  Regardless of the situation, we must always have at least one input (replay protection) and output (although outputs may be entirely change outputs).
  Then, depending on the situation, we may have 'withdrawal', 'reclaim' and or 'deposit' present.

  `reclaim` and `deposit` are "opposed" and can't be present together. `withdrawal` can in theory be present with both, but in practice, will only be present alone: we'll only
  allow users to withdraw their reward as part of standard transaction. `deposit` are only present when registering a stake-key for the first time. In summary, this gives us:

  type of transaction  | objects
  -------------------  | --------
  standard transaction | inputs, outputs, withdrawal? (if users said so)
  initial delegation   | inputs, outputs, deposit
  re-delegation        | inputs, outputs
  de-registration      | inputs, outputs, reclaim

- 227dcd9d9287d4ef3d76f602fc4cdd5b5e15d347
  :round_pushpin: **only count reclaim and withdrawal as part of the input balance if there's already a UTxO input**
  
- dd003064e32d82cdb53bfeb164ad3614c8b0fb93
  :round_pushpin: **tweak arbitrary coin selection generator to produce more realistic cases**
  The generator was good for 'payment' kind of selection. Yet, we also sort of abuse the fee balancing when creating selection for delegation with special delegation
with no inputs and no outputs. This was therefore completely untested by the current generator! Now it does, sometimes, create typical selections we would create when
delegating or undelegating. It nicely caught an error in the fee balancing algorithm which this commit also fixes

- 523ca66a8c5c7cc8417836830c4d467c29381ffa
  :round_pushpin: **re-implement shelley's transaction layer accordingly**
  In the end, it is much simpler and cleaner, no more hacky-hack to get things working. The newly introduced 'initDelegationSelection' makes it easier to construct a correct coin selection from the start and simply go with it.

- 7fa1162927b32e4510d93ae2c52753c2eae5ca55
  :round_pushpin: **review integration 'feeEstimator' to use 'DelegationAction'**
  We have two different ways of computing fees, and this is because in Byron, it used to be complicated to evaluate fees for a transaction. Now, this fee estimator feels sort of redundant and we should solely rely on the result from the fee estimation endpoint. In the meantime, I've adjusted it to make a bit more consistent with how fees are calculated elsewhere.

I've also adjusted the genesis file to have a much bigger deposit key.  The previous code was actually handling things in a very wrong way, but it went unnoticed because two errors were cancelling each others: fee were slightly over-evaluated, and the deposit for key was small enough to compensate.

- f8faf55502f57a203be4470954820df5b31fd8b3
  :round_pushpin: **re-implement byron & jormungandr transaction layers accordingly**
  
- ff7ee223cb549538025b75f6cd4cddccd3f3d7a7
  :round_pushpin: **add withdrawal on each payment transaction**
  We could make it slightly smarter here and requires that the withdrawal is only requested when the balance is somewhat sufficient.
Sufficient would mean, bigger than the cost of adding it. Indeed, adding a withdrawal would only make sense if the cost of adding it
isn't greater to its value.

- 98ace73a0f2e8d43faa0fa0859ea9ca2d2316def
  :round_pushpin: **only set withdrawal when they are sufficiently big**
  Otherwise, we end-up paying for more than the value they offer.

- d6b1db1038f2c455433dae5dc0b4dc079bc1b7fd
  :round_pushpin: **remove now obsolete 'ErrChangeIsEmptyForRetirement'**
  
- ba174deb629f70a9f19f6c406384f733543c05ca
  :round_pushpin: **prepare foundation to support withdrawals witnesses in standard transaction**
  I've once again changed a bit the transaction layer to be more aligned with the recent changes (now takes a full coin selection).
Also, it now requires a reward account to be given; this is relatively 'unsound' for Byron and Icarus wallet since they have no
reward account, yet, the account is only used when a non-null withdrawal is provided which should never happen for these wallet types
since they aren't receiving rewards at all.

- e7ab562ade96ec58d7d47aefd61da5c814164121
  :round_pushpin: **add property showing that withdrawals are correctly estimated by Shelley tx layer**
  
- 3035ec7dffe0ca0b63cb1185b73b5af0d1a19464
  :round_pushpin: **add integration test showing that rewards are spent with a transaction**
  
- 10e04f46d1d6efab8d6d5a74757c26aa31fd6c27
  :round_pushpin: **count rewards as part of the total and available balance**

# Comments

<!-- Additional comments or screenshots to attach if any -->

- ~~TODO: need to create the a corresponding withdrawal witness inside payment transactions~~
- ~~TODO: need to NOT add a withdrawal when it's below certain threshold~~
- ~~TODO: need to add an integration test showing that reward are withdrawn successfully.~~

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

KtorZ committed:
extend 'CoinSelection' type to carry an extra reserve amount that can be used as input

For now, the reserve is always 'Nothing', which should not alter any of the existing behavior. Then, it can be
set to a specific value (e.g. the reward balance) to be used on the left side of the balance (i.e. the input side).

KtorZ committed:
rename 'rebalanceChangeOutput' in 'rebalanceSelection' and allow picking from the reserve

The approach is still very 'simple', before trying to deplete any change output, we first try to remove fees
from the reserve. Once the reserve is empty, we'll start depleting change outputs.

KtorZ committed:
move prop_rebalanceSelection from byron package to core package

It was originally in `-byron` because it was using the fee policy and
fee estimation from Byron, so, while moving it, I also changed the way
the transaction size is estimated to make it mimics the way it's done on
shelley and byron, with rather realistic fee values.

KtorZ committed:
generate arbitrary selection with reserve, add more classification and assertions to properties

KtorZ commented on lib/byron/test/unit/Cardano/Wallet/Byron/TransactionSpec.hs:
```diff
@@ -157,17 +151,14 @@ spec = do
-    it "1561 - The fee balancing algorithm converges for any coin selection."
-        $ property
-        $ withMaxSuccess 10000
-        $ forAllBlind (genSelection @'Mainnet @ByronKey) prop_rebalanceChangeOutputs
```
Moved to core package.

KtorZ commented on lib/byron/test/unit/Cardano/Wallet/Byron/TransactionSpec.hs:
```diff
@@ -283,38 +274,6 @@ spec = do
-            newTransactionLayer @'Mainnet @ByronKey Proxy mainnetMagic
-        feePolicy =
-            LinearFee (Quantity 155381) (Quantity 43) (Quantity 0)
-
```
Moved to core package.

paweljakubas commented on lib/core/src/Cardano/Wallet/Primitive/CoinSelection.hs:
```diff
@@ -48,6 +48,8 @@ data CoinSelection = CoinSelection
       -- ^ Picked outputs
     , change  :: [Coin]
       -- ^ Resulting changes
+    , reserve :: Maybe Coin
```
it can be only "one" ? not list?

KtorZ replied:
See comments above. In theory, we could also have deposits here. So I am thinking about having a list of `Reserve` where `Reserve = Withdrawal | Deposit`

paweljakubas replied:
Will Withdraw and Deposit, if introduced will change on this level? Will there be different conditions in `rebalanceSelection` for them? for example Withdraw will deplete and transform into change at first, and only then Deposit?  Or other rules?

KtorZ replied:
No, they'll have the exact same effect hence why I initially went for a simple `Maybe Coin`. Perhaps it's not worth abstracting further here, in the end, both the deposit and withdrawal can be represent as a sum for this reserve. 
The difference lies in how the transaction is then constructed / signed / serialized. Withdrawals require a special entry as .. "withdrawals" in the transaction, as well as a special witness. 

Whereas deposit are implicit, when de-registering a stake key.

paweljakubas commented on lib/core/src/Cardano/Wallet/Primitive/Fee.hs:
```diff
@@ -213,16 +213,33 @@ coverRemainingFee (Fee fee) = go [] where
         (s, Fee 0)
 
+    -- some fee left to pay, and the reserve is non-empty, use it first.
+    | φ_original > δ_original && reserve s > Just (Coin 0) =
```
:+1:

paweljakubas commented on lib/core/src/Cardano/Wallet/Primitive/Fee.hs:
```diff
@@ -213,16 +213,33 @@ coverRemainingFee (Fee fee) = go [] where
+            -- Safe because of the above guard.
+            Just (Coin r) = reserve s
+            r' = if r > φ_original
+                then Coin (r - φ_original)
```
it was supposed to turn all reserve (if there is left) into change. So `rebalanceSelection` will be called in with decreased reserve and some branch will decrease it to zero and after that it will return in `senderPaysFee` with reserve (Just (Coin 0)). Is that right?

KtorZ replied:
Correct.

KtorZ replied:
Hence the added check in the corresponding property checking that the reserve is always 0 at the end of the balancing.

paweljakubas commented on lib/core/test/unit/Cardano/Wallet/Primitive/FeeSpec.hs:
```diff
@@ -472,6 +491,66 @@ propDivvyFeeInvariantEmptyList (fee, outs) =
+        & classify reserveLargerThanFee
+            "reserve > fee"
+        & classify feeLargerThanDelta
+            "fee > delta"
```
any chance to test `ErrValidateSelection` here or maybe at integration testing/backend TransactionSpecs ? :thinking:

KtorZ replied:
In the fee spec, no chance, and in integration also quite hard. This is actually a server error (500), so something that we shouldn't be able to trigger under any circumstances, in theory.

KtorZ committed:
do not mutate the reserve as part of the fee balancing process

Everything in the past coin selection (without reserve) was easy to reason about. One could always look at the sum
of inputs and compare it with the sum of outputs and changes, and there were some nice properties about this. Having
mutable reserve makes it somewhat very confusing because, there's no trace of where the money comes from (turning
reserve into changes shouldn't make it disappear).

KtorZ committed:
expand 'reserve' to also cover deposit and reclaim

So that types are clear from the coin selection and a bit less confusing.

KtorZ committed:
extend transaction layer with an 'initDelegationSelection'

  This in order to defer to the relevant backend the necessary bits of logic regarding withdrawal, deposits and reclaims.
  I've also removed the 'WalletDelegation' from 'mkDelegationJoinTx' as this can now be inferred from the presence of a deposit in the coin selection.

  We have the following situation:

  object     | cardano-node | jormungandr
   ----      | ----         | ----
  inputs     | explicit     | explicit
  reclaim    | implicit     | N/A
  withdrawal | explicit     | N/A
  ---        | ---          | ---
  outputs    | explicit     | explicit
  deposit    | implicit     | N/A

  Regardless of the situation, we must always have at least one input (replay protection) and output (although outputs may be entirely change outputs).
  Then, depending on the situation, we may have 'withdrawal', 'reclaim' and or 'deposit' present.

  `reclaim` and `deposit` are "opposed" and can't be present together. `withdrawal` can in theory be present with both, but in practice, will only be present alone: we'll only
  allow users to withdraw their reward as part of standard transaction. `deposit` are only present when registering a stake-key for the first time. In summary, this gives us:

  type of transaction  | objects
  -------------------  | --------
  standard transaction | inputs, outputs, withdrawal? (if users said so)
  initial delegation   | inputs, outputs, deposit
  re-delegation        | inputs, outputs
  de-registration      | inputs, outputs, reclaim

KtorZ committed:
only count reclaim and withdrawal as part of the input balance if there's already a UTxO input

KtorZ committed:
tweak arbitrary coin selection generator to produce more realistic cases

The generator was good for 'payment' kind of selection. Yet, we also sort of abuse the fee balancing when creating selection for delegation with special delegation
with no inputs and no outputs. This was therefore completely untested by the current generator! Now it does, sometimes, create typical selections we would create when
delegating or undelegating. It nicely caught an error in the fee balancing algorithm which this commit also fixes

KtorZ committed:
re-implement shelley's transaction layer accordingly

In the end, it is much simpler and cleaner, no more hacky-hack to get things working. The newly introduced 'initDelegationSelection' makes it easier to construct a correct coin selection from the start and simply go with it.

KtorZ committed:
review integration 'feeEstimator' to use 'DelegationAction'

We have two different ways of computing fees, and this is because in Byron, it used to be complicated to evaluate fees for a transaction. Now, this fee estimator feels sort of redundant and we should solely rely on the result from the fee estimation endpoint. In the meantime, I've adjusted it to make a bit more consistent with how fees are calculated elsewhere.

I've also adjusted the genesis file to have a much bigger deposit key.  The previous code was actually handling things in a very wrong way, but it went unnoticed because two errors were cancelling each others: fee were slightly over-evaluated, and the deposit for key was small enough to compensate.

KtorZ committed:
re-implement byron & jormungandr transaction layers accordingly

KtorZ committed:
add withdrawal on each payment transaction

We could make it slightly smarter here and requires that the withdrawal is only requested when the balance is somewhat sufficient.
Sufficient would mean, bigger than the cost of adding it. Indeed, adding a withdrawal would only make sense if the cost of adding it
isn't greater to its value.

paweljakubas commented on lib/core/src/Cardano/Wallet/Primitive/CoinSelection.hs:
```diff
@@ -44,10 +44,16 @@ import GHC.Generics
+    , withdrawal :: Word64
+      -- ^ An available withdrawal amount, counting as an extra input
+    , reclaim :: Word64
+      -- ^ Claim back a deposit, counting as a an extra input
```
:+1:

paweljakubas commented on lib/core/src/Cardano/Wallet/Transaction.hs:
```diff
@@ -97,30 +89,33 @@ data TransactionLayer t k = TransactionLayer
         -- The certificate is the public key of the reward account.
 
-    , minimumFee :: FeePolicy -> [Certificate] -> CoinSelection -> Fee
+    , initDelegationSelection
```
:+1:

paweljakubas commented on lib/core/test/unit/Cardano/Wallet/Primitive/FeeSpec.hs:
```diff
@@ -567,13 +663,58 @@ genTxOut coins = do
+genWithdrawal :: Gen Word64
+genWithdrawal = frequency
+    [ (3, pure 0)
+    , (1, oneof
```
why this oneof?

KtorZ replied:
To favor some edge-cases a bit more in the generator. The type is a `Word64`, so something like `arbitrary` will span over the entire domain. Yet, we know that bugs are usually found at edges, so I've favored small values, close to the fee amount so see how it behaves when a bit smaller than the fee, or a bit bigger, and still, allowed for big values far from this edge.

paweljakubas reviewed (approved):
A lot of changes in the code due to two ideas implemented (not all was able to follow, but I bet the author mainly followed compiler here). When reviewing focused on the ideas, ie. 1. extension of CoinSelection to account for deposit, withdrawal and reclaim, 2. changing interface in Cardano.Wallet.Transaction - and here using CoinSelection in number of methods instead of inputs/outputs and introducing `initDelegationSelection` to handle initial selection for different backends, 3. testing rebalancingSelection in unit tests. All three point looks solid to me!

KtorZ committed:
only set withdrawal when they are sufficiently big

Otherwise, we end-up paying for more than the value they offer.

KtorZ committed:
remove now obsolete 'ErrChangeIsEmptyForRetirement'

KtorZ committed:
prepare foundation to support withdrawals witnesses in standard transaction

I've once again changed a bit the transaction layer to be more aligned with the recent changes (now takes a full coin selection).
Also, it now requires a reward account to be given; this is relatively 'unsound' for Byron and Icarus wallet since they have no
reward account, yet, the account is only used when a non-null withdrawal is provided which should never happen for these wallet types
since they aren't receiving rewards at all.

KtorZ committed:
add property showing that withdrawals are correctly estimated by Shelley tx layer

KtorZ committed:
add integration test showing that rewards are spent with a transaction

KtorZ committed:
count rewards as part of the total and available balance
